### DISCUSSION TITLE
Quarto crashes when rendering

### DISCUSSION THREAD

**Discussion by Marlene283** [quarto-dev/quarto-cli#10994] [Q&A]
### Description

Hello,

I am new to github, so I am not sure if I post this question at the correct place.
Anyway:

I use quarto with R. I updated today R to version 4.4.1 and RStudio to 2024.09.0 Build 375. 
Since the updates, when I try to render a document (html output), quarto crashes with this output:

ERROR: Theme file compilation failed:

===== CRASH =====

Stack trace:

===== CRASH =====
    at dartCommand (file:///C:/PROGRA~ 1/RStudio/RESOUR~ 1/app/bin/quarto/bin/quarto.js:43445:15)
    at eventLoopTick (ext:core/01_core.js:153:7)
    at async dartCompile (file:///C:/PROGRA~ 1/RStudio/RESOUR~ 1/app/bin/quarto/bin/quarto.js:43407:5)
    at async compileWithCache (file:///C:/PROGRA~ 1/RStudio/RESOUR~ 1/app/bin/quarto/bin/quarto.js:43636:17)
    at async compileSass (file:///C:/PROGRA~ 1/RStudio/RESOUR~ 1/app/bin/quarto/bin/quarto.js:43497:12)
    at async resolveSassBundles (file:///C:/PROGRA~ 1/RStudio/RESOUR~ 1/app/bin/quarto/bin/quarto.js:65862:27)
    at async resolveExtras (file:///C:/PROGRA~ 1/RStudio/RESOUR~ 1/app/bin/quarto/bin/quarto.js:69904:18)
    at async runPandoc (file:///C:/PROGRA~ 1/RStudio/RESOUR~ 1/app/bin/quarto/bin/quarto.js:69425:24)
    at async renderPandoc (file:///C:/PROGRA~ 1/RStudio/RESOUR~ 1/app/bin/quarto/bin/quarto.js:70717:26)
    at async Object.onRender (file:///C:/PROGRA~ 1/RStudio/RESOUR~ 1/app/bin/quarto/bin/quarto.js:78268:36)
    
    

 When I use 'quarto check' in the terminal I get the following output:  

    C:\Users\Marlene\Documents>quarto check
Quarto 1.5.57
[>] Checking versions of quarto binary dependencies...
      Pandoc version 3.2.0: OK
      Dart Sass version 1.70.0: OK
      Deno version 1.41.0: OK
      Typst version 0.11.0: OK
[>] Checking versions of quarto dependencies......OK
[>] Checking Quarto installation......OK
      Version: 1.5.57
      Path: C:\Program Files\RStudio\resources\app\bin\quarto\bin
      CodePage: 1252

[>] Checking tools....................OK
      TinyTeX: (not installed)
      Chromium: (not installed)

[>] Checking LaTeX....................OK
      Tex:  (not detected)

[>] Checking basic markdown render....OK

ERROR: Theme file compilation failed:

===== CRASH =====

Stack trace:

===== CRASH =====
    at dartCommand (file:///C:/Program%20Files/RStudio/resources/app/bin/quarto/bin/quarto.js:43445:15)
    at eventLoopTick (ext:core/01_core.js:153:7)
    at async dartCompile (file:///C:/Program%20Files/RStudio/resources/app/bin/quarto/bin/quarto.js:43407:5)
    at async compileWithCache (file:///C:/Program%20Files/RStudio/resources/app/bin/quarto/bin/quarto.js:43636:17)
    at async compileSass (file:///C:/Program%20Files/RStudio/resources/app/bin/quarto/bin/quarto.js:43497:12)
    at async resolveSassBundles (file:///C:/Program%20Files/RStudio/resources/app/bin/quarto/bin/quarto.js:65862:27)
    at async resolveExtras (file:///C:/Program%20Files/RStudio/resources/app/bin/quarto/bin/quarto.js:69904:18)
    at async runPandoc (file:///C:/Program%20Files/RStudio/resources/app/bin/quarto/bin/quarto.js:69425:24)
    at async renderPandoc (file:///C:/Program%20Files/RStudio/resources/app/bin/quarto/bin/quarto.js:70717:26)
    at async Object.onRender (file:///C:/Program%20Files/RStudio/resources/app/bin/quarto/bin/quarto.js:78268:36)

I noticed the paths are a bit screwed up in the crash report (e.g. "Program Files" turned into "Program%20Files"), is that normal or might that be an issue?

I hope this covers everything you need to know about the issue. I am not sure what to do, this exceeds my skills in error fixing. Can you help me?

Thanks for your help!

**Comment by cscheid** [comment]
The fact that `quarto check` itself is failing is worrisome. @cderv What do you think?

**Reply by cderv** [reply]
It could be related to this one 
- https://github.com/quarto-dev/quarto-cli/issues/6651

Dart for windows is a BAT file, and we still execute this. #6651 has not yet been dealt with (we discussed replacing the bat file they provide). 

In some environment it causes issues. But #6651 fails even at getting the dart version, which seems to be different here. 

We are supposed to catch the error, but it seems it is empty. 

@Marlene283 can you activate verbose debug mode ? 
See https://quarto.org/docs/troubleshooting/#verbose-mode

and then post the new logging you have. 

Thank you !

**Reply by Marlene283** [reply]
Then I get this output:

Quarto version: 1.5.57
ERROR: No valid input files passed to render

Stack trace:
    at Command.actionHandler (file:///C:/PROGRA~1/RStudio/RESOUR~1/app/bin/quarto/bin/quarto.js:83100:15)
    at eventLoopTick (ext:core/01_core.js:153:7)
    at async Command.execute (file:///C:/PROGRA~1/RStudio/RESOUR~1/app/bin/quarto/bin/quarto.js:8017:13)
    at async Command.parseCommand (file:///C:/PROGRA~1/RStudio/RESOUR~1/app/bin/quarto/bin/quarto.js:7907:20)
    at async quarto (file:///C:/PROGRA~1/RStudio/RESOUR~1/app/bin/quarto/bin/quarto.js:118224:9)
    at async file:///C:/PROGRA~1/RStudio/RESOUR~1/app/bin/quarto/bin/quarto.js:118244:9
    at async mainRunner (file:///C:/PROGRA~1/RStudio/RESOUR~1/app/bin/quarto/bin/quarto.js:118128:9)
    at async file:///C:/PROGRA~1/RStudio/RESOUR~1/app/bin/quarto/bin/quarto.js:118235:5
Error in `quarto::quarto_render()`:
✖ Error running quarto cli.
Caused by error:
! System command 'quarto.exe' failed
Run `rlang::last_trace()` to see where the error occurred.

**Reply by cderv** [reply]
> Error in quarto::quarto_render():

Can you just try using `quarto check` in a Terminal, and not using R layer that is calling quarto. 

thanks

**Reply by Marlene283** [reply]
Sorry, I am not sure, I understand what you mean. Do you mean in the cmd terminal for example? If I try using it there, I have to set the path to C:\Program Files\RStudio\resources\app\bin\quarto\bin before I can use it and that just gives me the same output as the one described above, in the description of my problem. My apologies, if I am doing the wrong thing here.

Thank you so much for your help

**Reply by cderv** [reply]
Ok thanks. You are in RStudio IDE then. 

Try this 
1. In Rstudio IDE terminal, if not already in Powershell terminal call Powershell (`Powershell` or `pwsh` if you have it
2. Then set the environment for debug log 
````powershell
$env:QUARTO_LOG_LEVEL="DEBUG"
````
3. then call `quarto check` - `quarto` command should be found as you are in RStudio
	- To do this outside of rstudio ide, call the quarto binary with its full path. 


What I am interested in is the more verbose output before the error happens. 

Thanks

**Reply by Marlene283** [reply]
This then returns:

Quarto version: 1.5.57
Quarto 1.5.57
[>] Checking versions of quarto binary dependencies...
[execProcess] C:\Program Files\RStudio\resources\app\bin\quarto\bin\tools\pandoc --version
[execProcess] Success: true, code: 0
      Pandoc version 3.2.0: OK
[execProcess] C:\Program Files\RStudio\resources\app\bin\quarto\bin\tools\x86_64\dart-sass\sass.bat --version
[execProcess] Success: true, code: 0
      Dart Sass version 1.70.0: OK
      Deno version 1.41.0: OK
[execProcess] C:\Program Files\RStudio\resources\app\bin\quarto\bin\tools\x86_64\typst --version
[execProcess] Success: true, code: 0
      Typst version 0.11.0: OK
[>] Checking versions of quarto dependencies......OK
[>] Checking Quarto installation......OK
      Version: 1.5.57
      Path: C:\Program Files\RStudio\resources\app\bin\quarto\bin
[execProcess] reg query HKLM\SYSTEM\CurrentControlSet\Control\Nls\CodePage /v ACP
[execProcess] Success: true, code: 0
      CodePage: 1252

[execProcess] cmd /c C:\Users\Marlene\AppData\Local\Temp\quarto-safe-exec52cc9e36274e2c84.bat
[execProcess] Success: false, code: 1
[>] Checking tools....................OK
      TinyTeX: (not installed)
      Chromium: (not installed)

[execProcess] cmd /c C:\Users\Marlene\AppData\Local\Temp\quarto-safe-execf832b41b345d0dcf.bat
[execProcess] Success: false, code: 1
[execProcess] cmd /c C:\Users\Marlene\AppData\Local\Temp\quarto-safe-exec19214a25930493b8.bat
[execProcess] Success: false, code: 1
[>] Checking LaTeX....................OK
      Tex:  (not detected)

Loaded deno-dom-native
[execProcess] C:\Program Files\RStudio\resources\app\bin\quarto\bin\tools\x86_64\dart-sass\sass.bat C:\Users\Marlene\AppData\Local\Temp\quarto-session8148612c128e0702\da757819c0ecdb89\5213e0ba42faa4e9.scss C:\Users\Marlene\AppData\Local\Temp\quarto-session8148612c128e0702\da757819c0ecdb89\sass\397ef2e52d54cf686e4908b90039e9db.css --style compressed --quiet --load-path=C:\Program Files\RStudio\resources\app\bin\quarto\share\formats\html\bootstrap\dist\scss --load-path=C:\Program Files\RStudio\resources\app\bin\quarto\share\formats\html\bslib\bslib-scss
(|) Checking basic markdown render....[execProcess] Success: false, code: -1073740791
[DART path]   : C:\Program Files\RStudio\resources\app\bin\quarto\bin\tools\x86_64\dart-sass\sass.bat
[DART args]   : C:\Users\Marlene\AppData\Local\Temp\quarto-session8148612c128e0702\da757819c0ecdb89\5213e0ba42faa4e9.scss C:\Users\Marlene\AppData\Local\Temp\quarto-session8148612c128e0702\da757819c0ecdb89\sass\397ef2e52d54cf686e4908b90039e9db.css --style compressed --quiet --load-path=C:\Program Files\RStudio\resources\app\bin\quarto\share\formats\html\bootstrap\dist\scss --load-path=C:\Program Files\RStudio\resources\app\bin\quarto\share\formats\html\bslib\bslib-scss
[DART stdout] :
[DART stderr] :
===== CRASH =====
ExceptionCode=-1073741795, ExceptionFlags=0, ExceptionAddress=000002136819F74C

[>] Checking basic markdown render....OK

[NotebookContext]: Starting Cleanup
ERROR: Theme file compilation failed:


===== CRASH =====

Stack trace:


===== CRASH =====
    at dartCommand (file:///C:/Program%20Files/RStudio/resources/app/bin/quarto/bin/quarto.js:43445:15)
    at eventLoopTick (ext:core/01_core.js:153:7)
    at async dartCompile (file:///C:/Program%20Files/RStudio/resources/app/bin/quarto/bin/quarto.js:43407:5)
    at async compileWithCache (file:///C:/Program%20Files/RStudio/resources/app/bin/quarto/bin/quarto.js:43636:17)
    at async compileSass (file:///C:/Program%20Files/RStudio/resources/app/bin/quarto/bin/quarto.js:43497:12)
    at async resolveSassBundles (file:///C:/Program%20Files/RStudio/resources/app/bin/quarto/bin/quarto.js:65862:27)
    at async resolveExtras (file:///C:/Program%20Files/RStudio/resources/app/bin/quarto/bin/quarto.js:69904:18)
    at async runPandoc (file:///C:/Program%20Files/RStudio/resources/app/bin/quarto/bin/quarto.js:69425:24)
    at async renderPandoc (file:///C:/Program%20Files/RStudio/resources/app/bin/quarto/bin/quarto.js:70717:26)
    at async Object.onRender (file:///C:/Program%20Files/RStudio/resources/app/bin/quarto/bin/quarto.js:78268:36)


Does that help you? Thanks for your patience!

**Reply by cscheid** [reply]
I think @cderv's suspicion about .bat files is accurate:

```
[execProcess] cmd /c C:\Users\Marlene\AppData\Local\Temp\quarto-safe-exec52cc9e36274e2c84.bat
[execProcess] Success: false, code: 1
...
[execProcess] cmd /c C:\Users\Marlene\AppData\Local\Temp\quarto-safe-execf832b41b345d0dcf.bat
[execProcess] Success: false, code: 1
[execProcess] cmd /c C:\Users\Marlene\AppData\Local\Temp\quarto-safe-exec19214a25930493b8.bat
[execProcess] Success: false, code: 1
```

All of these come from:

https://github.com/quarto-dev/quarto-cli/blob/740829733499855335a7861e3c4f1333a09db45d/src/core/windows.ts#L145-L162

Christophe, the relevant commit is this: https://github.com/quarto-dev/quarto-cli/commit/99d6d9a8059e60cf40073c699a644ead5fc391bc

It mentions a "deno issue". I wonder if this is now solved.

**Reply by cderv** [reply]
Yes this is related to #6651 

I believe there is restriction on this environment about executing .bat file. 

We do use `.bat` file to handle the quoting problem we encountered by running deno.run on windows. 
Maybe Deno.Command() does not have the issue... It is hard to test - it initially came with running `tlmgr.bat` from Deno itself. (yes tlmgr is a BAT file - and we needed to wrap it in a .bat file to avoid quoting issue on windows - they are painful!)

Though we don't wrap the call to Dart sass. It is a `saas.bat` which is executed and it seems it is not succeeding either 

````
[execProcess] C:\Program Files\RStudio\resources\app\bin\quarto\bin\tools\x86_64\dart-sass\sass.bat C:\Users\Marlene\AppData\Local\Temp\quarto-session8148612c128e0702\da757819c0ecdb89\5213e0ba42faa4e9.scss C:\Users\Marlene\AppData\Local\Temp\quarto-session8148612c128e0702\da757819c0ecdb89\sass\397ef2e52d54cf686e4908b90039e9db.css --style compressed --quiet --load-path=C:\Program Files\RStudio\resources\app\bin\quarto\share\formats\html\bootstrap\dist\scss --load-path=C:\Program Files\RStudio\resources\app\bin\quarto\share\formats\html\bslib\bslib-scss
(|) Checking basic markdown render....[execProcess] Success: false, code: -1073740791
````

This is not related to our `safeWindowsExec()`. So I still think we are in similar case than #6651 - meaning in an environment where .bat file are not considered safe to run IMO. 

Weird thing is that version check worked :thinking:

````
[execProcess] C:\Program Files\RStudio\resources\app\bin\quarto\bin\tools\x86_64\dart-sass\sass.bat --version
[execProcess] Success: true, code: 0
````

Also about 
````
===== CRASH =====
ExceptionCode=-1073741795, ExceptionFlags=0, ExceptionAddress=000002136819F74C
````

I do think `-1073741795` error code is related on windows to Administration Access limitation. Something seems to trigger User Access Control on Windows, and this environment is blocked - meaning it errors

@Marlene283 do you know if you have restriction on .bat file ? If you don't know what I am talking about - really no issue. Just asking in case you know.

**Reply by Marlene283** [reply]
So first, yes I am not really sure how to check. But I asked ChatGPT how to, which gave me a couple of things to check. Since I am not sure what info you need, I just tell you all:

1. Creating a .bat file with this code:

@echo off
echo Test .bat file is working
pause

and executing it, worked.


2. I checked with the Local Group Polcy Editor and there are no policies enabled (not configurated).

3. The antivirus software does not block any specific file types.

4. Checking in powershell with:
PS C:\WINDOWS\system32> Get-ExecutionPolicy
Restricted
PS C:\WINDOWS\system32> Get-ExecutionPolicy -List

        Scope ExecutionPolicy
        ----- ---------------
MachinePolicy       Undefined
   UserPolicy       Undefined
      Process       Undefined
  CurrentUser       Undefined
 LocalMachine       Undefined


So it seems nothing is restriced here, right? If I should check in a another different way, please tell me. Thanks for your help

**Reply by Marlene283** [reply]
@cderv @cscheid When I started to have the issue last week, I installed the newest verion of R and Rstudio probably like 5 times in slightly different ways. I also tried a older version of R, which did not resolve the issue. The only thing I forgot to try out, was trying a older version of RStudio. I tried this today and with the 2024.04.2 Build 764 version, I don't have the issue anymore

**Reply by cderv** [reply]
> The only thing I forgot to try out, was trying a older version of RStudio. I tried this today and with the 2024.04.2 Build 764 version, I don't have the issue anymore

This is interesting ! Only thing I can see related to Quarto is that it is using a different version 🤔 

Do you know which version you have know ? `quarto check` inside RStudio IDE terminal now returns what ? 

I really would like to understand the problem here... 🤔

**Reply by Marlene283** [reply]
C:\Users\Marlene\Documents>quarto check
Quarto 1.4.555
[>] Checking versions of quarto binary dependencies...
      Pandoc version 3.1.11: OK
      Dart Sass version 1.69.5: OK
      Deno version 1.37.2: OK
[>] Checking versions of quarto dependencies......OK
[>] Checking Quarto installation......OK
      Version: 1.4.555
      Path: C:\Program Files\RStudio\resources\app\bin\quarto\bin
      CodePage: 1252

[>] Checking tools....................OK
      TinyTeX: (not installed)
      Chromium: (not installed)

[>] Checking LaTeX....................OK
      Tex:  (not detected)

[>] Checking basic markdown render....OK

[>] Checking Python 3 installation....(None)

      Unable to locate an installed version of Python 3.
      Install Python 3 from https://www.python.org/downloads/

[>] Checking R installation...........OK
      Version: 4.4.1
      Path: C:/PROGRA~1/R/R-44~1.1
      LibPaths:
        - C:/Users/Marlene/AppData/Local/R/win-library/4.4
        - C:/Program Files/R/R-4.4.1/library
      knitr: 1.48
      rmarkdown: 2.28

[>] Checking Knitr engine render......OK

**Reply by cderv** [reply]
So quarto 1.4.555 seems to work for you and latest 1.5.57 does not 😞 

as I can't reproduce this won't be easy to understand what is the problem... 

If you are willing to test more to help us, you could try install quarto 1.5.57 without updating RStudio. Just download and install from https://quarto.org/docs/download/ and see if the problems happens again. 

Maybe you system flag the older binaries wrongly...

In the meantime, I'll see the changes that happened between those version. 
This rules out previous commit mentioned at https://github.com/quarto-dev/quarto-cli/discussions/10994#discussioncomment-10882783 as it is older.

**Reply by Marlene283** [reply]
I tried it out and with only updating quarto to version 1.5.57 without updating RStudio I do still get the same problem. Reverting back to quarto 1.4.555 works again

**Reply by cderv** [reply]
Thanks a lot for trying this ! We have narrow down a little which may help us find the problem, even if not reproducible. It could also be an update in dependencies between 1.4.555 and 1.5.57 but this is more problematic. 

I really wish I can understand how to reproduce
